### PR TITLE
gh-92123: Convert `_elementtree` types to heap types

### PR DESCRIPTION
**Pull request by erlend-aasland**
GH-23428 resurrected.

* Issue: gh-92123

### CODE REVIEW

**Comment by kumaraditya303** [comment]
With https://github.com/python/cpython/issues/100689 the capsule also needs to isolated. LMK when this goes out of draft.

**Comment by erlend-aasland** [comment]
> With https://github.com/python/cpython/issues/100689 the capsule also needs to isolated. LMK when this goes out of draft. 

Yes, I also made a note of that. I hope to be able to allocate time for CPython this weekend.

**Comment by erlend-aasland** [comment]
@kumaraditya303: I wonder if we should tie this PR to gh-92123, or if we should create a targeted issue specifically for isolating `_elementtree`.

File: Modules/_elementtree.c [hunk 1]
```diff
@@ -4174,46 +4134,22 @@ static PyMethodDef treebuilder_methods[] = {
...
+    {Py_tp_init, _elementtree_TreeBuilder___init__},
+    {Py_tp_alloc, PyType_GenericAlloc},
+    {Py_tp_new, treebuilder_new},
+    {0, NULL},
+};
+
+static PyType_Spec treebuilder_spec = {
+    .name = "xml.etree.ElementTree.TreeBuilder",
+    .basicsize = sizeof(TreeBuilderObject),
+    .flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE | Py_TPFLAGS_HAVE_GC,
```
Comment (kumaraditya303) [inline Modules/_elementtree.c]: Same

File: Modules/_elementtree.c [hunk 2]
```diff
@@ -4224,46 +4160,24 @@ static PyMethodDef xmlparser_methods[] = {
...
+    {Py_tp_init, _elementtree_XMLParser___init__},
+    {Py_tp_alloc, PyType_GenericAlloc},
+    {Py_tp_new, xmlparser_new},
+    {0, NULL},
+};
+
+static PyType_Spec xmlparser_spec = {
+    .name = "xml.etree.ElementTree.XMLParser",
+    .basicsize = sizeof(XMLParserObject),
+    .flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE | Py_TPFLAGS_HAVE_GC,
```
Comment (kumaraditya303) [inline Modules/_elementtree.c]: Add `Py_TPFLAGS_IMMUTABLETYPE`.

File: Modules/_elementtree.c [hunk 3]
```diff
@@ -4122,46 +4096,32 @@ static PyGetSetDef element_getsetlist[] = {
...
+    {Py_mp_length, element_length},
+    {Py_mp_subscript, element_subscr},
+    {Py_mp_ass_subscript, element_ass_subscr},
+    {0, NULL},
+};
+
+static PyType_Spec element_spec = {
+    .name = "xml.etree.ElementTree.Element",
+    .basicsize = sizeof(ElementObject),
+    .flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE | Py_TPFLAGS_HAVE_GC,
```
Comment (kumaraditya303) [inline Modules/_elementtree.c]: Same

File: Modules/_elementtree.c [hunk 4]
```diff
@@ -2183,49 +2180,21 @@ elementiter_next(ElementIterObject *it)
...
-    0,                                          /* tp_dict */
-    0,                                          /* tp_descr_get */
-    0,                                          /* tp_descr_set */
-    0,                                          /* tp_dictoffset */
-    0,                                          /* tp_init */
-    0,                                          /* tp_alloc */
-    0,                                          /* tp_new */
+    .name = "_elementtree._element_iterator",
+    .basicsize = sizeof(ElementIterObject),
+    .flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
```
Comment (kumaraditya303) [inline Modules/_elementtree.c]: Same

**Comment by erlend-aasland** [comment]
@kumaraditya303: thanks for the immutable types reminder; I also had forgotten to disallow instantiation for the iter type.

File: Modules/_elementtree.c [hunk 5]
```diff
@@ -4334,10 +4258,10 @@ PyInit__elementtree(void)
...
                             "pyexpat version is incompatible");
-            return NULL;
+            goto error;
         }
     } else {
-        return NULL;
+        goto error;
     }
 
     st->str_append = PyUnicode_InternFromString("append");
```
Comment (kumaraditya303) [inline Modules/_elementtree.c:4267]: Also change `NULL` checks like on line 4269 to `goto error`?

Comment (erlend-aasland) [reply Modules/_elementtree.c:4267]: Ah, that a result of a bad merge. Thanks!

Comment (erlend-aasland) [reply Modules/_elementtree.c:4267]: BTW, the interned strings are not freed in the module clear/free functions.

**Review by kumaraditya303 — APPROVED** [review]
LGTM

**Comment by erlend-aasland** [comment]
All right, let's land this; I'll open PRs for getting rid of `ET_STATE_GLOBAL` right away.

**Comment by erlend-aasland** [comment]
Thanks for the great review, Kumar!

**Pull request merged**